File: LeetCode75_Artur/Sliding Window/1004_Max_Consecutive_OnesIII.py
"""
 https://leetcode.com/problems/max-consecutive-ones-iii/description/?envType=study-plan-v2&envId=leetcode-75
"""
from typing import Optional, List

class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        left, right = 0, 0
        num_zeros, num_ones = 0, 0
        max_ones = 0
        
        while right < len(nums):                
            if nums[right]:
                num_ones += 1
            else:
                num_zeros += 1
                
            while num_zeros > k:
                if nums[left]:
                    num_ones -= 1
                else:
                    num_zeros -= 1
                left += 1
            
            max_ones = max(max_ones, num_ones + num_zeros)
            right += 1
        
        return max_ones

    
if __name__ == "__main__":
    test1 = [1,1,1,0,0,0,1,1,1,1,0]
    test2 = [0,0,1,1,0,0,1,1,1,0,1,1,0,0,0,1,1,1,1]
    
    print(Solution().longestOnes(test1, 2))
    print(Solution().longestOnes(test2, 3))

File: LeetCode75_Artur/Sliding Window/1004_Max_Consecutive_OnesIII.cpp
/*
 https://leetcode.com/problems/max-consecutive-ones-iii/description/?envType=study-plan-v2&envId=leetcode-75
*/

#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    int longestOnes(vector<int>& nums, int k)
    {
        int numZeros = 0;
        int numOnes = 0;
        int maxOnes = 0;

        int left = 0;
        int right = 0;
        while( right < nums.size() )
        {
            if( nums[right++] )
                numOnes++;
            else
                numZeros++;

            while( numZeros > k )
            {
                if( nums[left] )
                    numOnes--;
                else
                    numZeros--;
                left++;
            }

            maxOnes = max(maxOnes, numOnes + numZeros);
        }

        return maxOnes;
    }
};


int main()
{
    Solution solution;

    vector<int> test1 = {1,1,1,0,0,0,1,1,1,1,0};
    vector<int> test2 = {0,0,1,1,0,0,1,1,1,0,1,1,0,0,0,1,1,1,1};

    cout << solution.longestOnes(test1, 2) << endl;
    cout << solution.longestOnes(test2, 3) << endl;
}

/*
 Main idea:
 Є два поінтери, які обмежують підпослідовність, яка
 має щонайбільше k нулів. Якщо нулів є більше за k,
 то сунемо лівий поінтер до тих пір, поки нулів не буде
 рівно k.

 Рахуємо кількість нулів та одиниць у підпослідовності,
 якщо нулів є щонайбільше k, то порівнюємо максимальну
 кількість із кількістю елементів у цій підпослідовності.

 Часова складність алгоритму: O(n)
*/